executor/stale_txn_test.go

```diff
@@ -1106,3 +1106,37 @@ func (s *testStaleTxnSuite) TestStmtCtxStaleFlag(c *C) {
 		c.Assert(tk.Se.GetSessionVars().StmtCtx.IsStaleness, IsFalse)
 	}
 }
+
+func (s *testStaleTxnSerialSuite) TestStaleSessionQuery(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	// For mocktikv, safe point is not initialized, we manually insert it for snapshot to use.
+	safePointName := "tikv_gc_safe_point"
+	safePointValue := "20160102-15:04:05 -0700"
+	safePointComment := "All versions after safe point can be accessed. (DO NOT EDIT)"
+	updateSafePoint := fmt.Sprintf(`INSERT INTO mysql.tidb VALUES ('%[1]s', '%[2]s', '%[3]s')
+	ON DUPLICATE KEY
+	UPDATE variable_value = '%[2]s', comment = '%[3]s'`, safePointName, safePointValue, safePointComment)
+	tk.MustExec(updateSafePoint)
+	tk.MustExec("use test")
+	tk.MustExec("create table t10 (id int);")
+	time.Sleep(3 * time.Second)
+	now := time.Now()
+	tk.MustExec(`set @@tidb_read_staleness="-2s"`)
+	// query will use staleness read
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/expression/injectNow", fmt.Sprintf(`return(%d)`, now.Unix())), IsNil)
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/executor/assertStaleTSO", fmt.Sprintf(`return(%d)`, now.Unix()-2)), IsNil)
+	tk.MustExec("select * from t10;")
+	c.Assert(failpoint.Disable("github.com/pingcap/tidb/executor/assertStaleTSO"), IsNil)
+	c.Assert(failpoint.Disable("github.com/pingcap/tidb/expression/injectNow"), IsNil)
+	// begin transaction won't be affected by read staleness
+	tk.MustExec("begin")
+	tk.MustExec("insert into t10(id) values (1);")
+	tk.MustExec("commit")
+	tk.MustExec("insert into t10(id) values (2);")
+	// query will still use staleness read
```

Comment:
What about adding a case that we use `as of timestamp...` after setting `@@tidb_read_staleness`?

Reply:
The compatibility between different cases of stale read will be pushed in next request.